test/test-utils.mjs

```diff
@@ -13,24 +13,96 @@
 // limitations under the License.
 
 import chalk from 'chalk'
+import {fileURLToPath} from 'node:url'
+import {relative} from 'node:path'
+import {sleep} from '../src/index.mjs'
 
 export {strict as assert} from 'assert'
 
-let всегоТестов = 0
+let queued = 0
+let passed = 0
+let failed = 0
+let total = 0
+let skipped = 0
+let focused = 0
 
-export function test(group, name) {
-  let фильтр = process.argv[3] || '.'
-  if (RegExp(фильтр).test(name) || RegExp(фильтр).test(group)) {
-    console.log('\n' + chalk.bgGreenBright.black(`${chalk.inverse(' ' + group + ' ')} ${name} `))
-    всегоТестов++
-    return true
+const singleThread = (fn) => {
+  let p = Promise.resolve()
+  return async function (...args) {
+    return (p = p.catch(_ => _).then(() => fn.call(this, ...args)))
   }
-  return false
 }
 
+const run = singleThread((cb) => timeout(cb(), 5000))
+
+const warmup = sleep(100)
+
+const timeout = (promise, ms, exception = `TimeoutException: exec time exceeds ${ms}ms`) => {
+  let timer
+  return Promise.race([
+    promise,
+    new Promise((_r, rej) => timer = setTimeout(rej, ms, exception))
+  ]).finally(() => clearTimeout(timer))
+}
+
+const log = (name, group, err, file = '') => {
+  if (err) {
+    console.log(err)
+    console.log(file)
+  }
+  console.log('\n' + chalk[ err ? 'bgRedBright' : 'bgGreenBright' ].black(`${chalk.inverse(' ' + group + ' ')} ${name} `))
+}
+
+export const test = async function (name, cb, focus, skip) {
+  const filter = RegExp(process.argv[3] || '.')
+  const {group, meta} = this
+  const file = meta ? relative(process.cwd(), fileURLToPath(meta.url)) : ''
+
+  if (filter.test(name) || filter.test(group) || filter.test(file)) {
+    focused += +!!focus
+    queued++
+
+    await warmup
```

Comment:
Why this warmup needed?

Reply:
There's no test files loader (like in jest), so we don't know how many files and how many test directives (only, skip) there will be at all. `ava/uvu` uses `test.run()` for this purpose. I think a short delay is better for our case.

Reply:
I’m still not getting it. Logic with timeouts seems over complicated. Maybe remove this part and don’t introduce timeouts for tests?

Reply:
Here's an example:
```js
// some.test.js
test(...)
test(...)
// one-more.test.js
test(...)
test.only()
```
We need to check, that there's no `only` in the entire test suite _before_ running the first test.

Reply:
I see. Isn’t it too much? I have doubts about implementing complicated testing framework.

Reply:
Race here: resolve test callback promise or throw timeout exception.
```js
const timeout = (promise, ms, exception = `TimeoutException: exec time exceeds ${ms}ms`) => {
  let timer
  return Promise.race([
    promise,
    new Promise((_r, rej) => timer = setTimeout(rej, ms, exception))
  ]).finally(() => clearTimeout(timer))
}
```

Reply:
I see. What test use this timeout functionality?

Reply:
All of them: 
```js
const run = singleThread((cb) => timeout(cb(), 5000)).
```
Default timeout for every test. Like in jest.

Reply:
But I why do we need timeouts in a first place?

Reply:
Just to highlight problems: If some test exceeds 5000 ms, it may indicate an implementation issue.

Reply:
Try, this will lead to "stuck" GitHub action, as we saw =) Let's discuss this tomorrow?

Reply:
Deal. I'm also available for tg / zoom / emails after 18:00 (UTC +3).

Reply:
* I've introduced two levels of protection agaist hanging: this one race-wrapped timeout and `- run: timeout 30 npm test` inside the test action.
* I can drop `warmup` if test src loading goes to test-utils too

Reply:
I still not convinced we need Promise.race stuff. Can we solve it via GitHub actions timeout?

Reply:
Ok, removed this part.

Reply:
Looks like still here. Also, is timeout available on windows? May switch to GH action one?

Reply:
* The diff: https://github.com/google/zx/pull/360/commits/220c8a6522714c2d3ed3d45af1a6b7a7551bbc49
* `timeout` for Windows:
  * https://superuser.com/questions/520527/windows-command-line-time-limit
  * https://stackoverflow.com/questions/46374893/timeout-equivalent-for-windows
  
I found [`timeout-cli`](https://github.com/anaisbetts/timeout-cli) pkg, but it's definitely overcomlicated inside. How about just `npx zx --experimental <<< 'withTimeout(30000)`npm run test`'`?

Reply:
I can add windows to the test matrix later.

Reply:
Why not just use https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes ?

Reply:
Finally pushed.

Reply:
btw, I've landed this `test-utils.js` snippet to [my pet project](https://github.com/antongolub/npm-registry-firewall/blob/master/src/test/test-utils.js), it works fine.